server/src/main/java/org/opensearch/gateway/BaseGatewayShardAllocator.java

```diff
@@ -78,6 +81,19 @@ public void allocateUnassigned(
         executeDecision(shardRouting, allocateUnassignedDecision, allocation, unassignedAllocationHandler);
     }
 
+    protected void allocateUnassignedBatchOnTimeout(List<ShardRouting> shardRoutings, RoutingAllocation allocation, boolean primary) {
+        Set<ShardRouting> batchShardRoutingSet = new HashSet<>(shardRoutings);
```

Comment:
Equals comparison based on `ShardRouting` for shards might not match all the replicas (as unassignedInfo could differ). It is better to construct the set based on ShardId

Reply:
+1, i think it was like that in first revision. After removing code duplication, it was missed. @imRishN

Reply:
Pushed fix